[webkitdirs] Compute Xcode destination specifier

emw-apple opened the pull request:
#### a775f3c7a4e4ebb6da8b828b53f9342a05640943
<pre>
[webkitdirs] Compute Xcode destination specifier
<a href="https://bugs.webkit.org/show_bug.cgi?id=266493">https://bugs.webkit.org/show_bug.cgi?id=266493</a>
<a href="https://rdar.apple.com/86340301">rdar://86340301</a>

Reviewed by Alexey Proskuryakov.

Instead of letting xcodebuild pick an ambiguous destination for us, have
webkitdirs compute a destination string based on the SDK and
architecture(s). This allows us to remove much of the hardcoded
architecture logic in webkitdirs, letting the build system figure out
which architecture(s) to use on its own.

Some nice side effects of this change:

- It is possible to build for a non-Mac platform just by setting
  SDKROOT. No need to override ARCHS.

- When building for macOS or a simulated platform, the default behaviors
  correspond to destinations in Xcode&apos;s destination menu. Building with
  no architecture override selects the &quot;My Mac&quot; destination or a
  simulator on the host system. Building with multiple architectures
  selects one of the &quot;Any &lt;platform&gt; Device|Simulator&quot; destinations.
  Aligning the destinations in this way helps keep incremental builds
  functional when switching between Make and Xcode.

To simplify the problem space, remove set-webkit-configuration&apos;s ability
to adjust the architecture, and webkitdirs&apos; corresponding knowledge of
&quot;passed&quot; architecture. set-webkit-configuration&apos;s architecture flags
have not been updated since Apple&apos;s 64-bit transition in 2009
(<a href="https://commits.webkit.org/33755@main).">https://commits.webkit.org/33755@main).</a>

* Makefile.shared: Pass the active SDK and arch to webkitdirs, to
  compute the destination string. Work around shell backticks being
  unable to expand an argument containing spaces with a $(shell ...)
  directive. Since the variable is only expanded once, webkitdirs is
  only called once.

* Tools/Scripts/build-imagediff: No longer need to override architecture.

* Tools/Scripts/set-webkit-configuration: Remove --32-bit and --64-bit
  flags which save state in the build directory. This is an old
  mechanism from before Apple&apos;s 64-bit transition, and is not used by
  the present day 32-bit WPE configurations.
(printCurrentSettings):

* Tools/Scripts/webkitdirs.pm:
(determineArchitecture): Don&apos;t look for set-webkit-configuration&apos;s state
  file.
(determineDestination): Added, computes an xcodebuild destination string
  based on SDK and architecture(s).
(determineXcodeSDKPlatformName): Refactor to allow a &quot;maccatalyst&quot;
  platform name when a SDK is also provided.
(determineXcodeSDK): Special case for &quot;maccatalyst&quot; as a platform name.
(XcodeOptions):
(XcodeOptionString): Enclose arguments containing spaces in quotes, for
  shell substitution.
(setConfiguration):
(destination): Added.
(determinePassedArchitecture): Deleted.
(passedArchitecture): Deleted.
(setArchitecture): Deleted.

* Tools/Scripts/webkitperl/BuildSubproject.pm: Argument parser was
  relying on shouldBuild32Bit() being called early on in evaluation,
  make the call explicit.

Canonical link: <a href="https://commits.webkit.org/273153@main">https://commits.webkit.org/273153@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fd1a206e9ef561d4e12fde1718548862e8bdb072

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/34464 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/13283 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/36466 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/37151 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/31165 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/35560 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/15690 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/10391 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/30157 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/34985 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/11248 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/30731 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/9823 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/9913 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/30787 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/38429 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/29292 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/31311 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/31103 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/35972 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/34352 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/10025 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/7911 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/33920 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/11842 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/40932 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7923 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/10581 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/8549 "Passed tests") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/10885 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

emw-apple commented on Tools/Scripts/webkitdirs.pm:
```diff
@@ -1284,9 +1324,6 @@ sub XcodeOptions
     push @options, "WK_LTO_MODE=$ltoMode" if $ltoMode;
     push @options, @baseProductDirOption;
-    push @options, "ARCHS=$architecture" if $architecture;
-    push @options, "ONLY_ACTIVE_ARCH=NO" if $didUserSpecifyArchitecture;
```
Note to self, we should do something about `ONLY_ACTIVE_ARCH`. When passed, it invalidates incremental builds (because it changes script phase environment variables). And it's a no-op when building to a generic destination, since those have no concept of an "active" architecture.

aproskuryakov commented on Tools/Scripts/webkitdirs.pm:
```diff
@@ -553,31 +553,13 @@ sub determineArchitecture
     if (isAppleCocoaWebKit() && $architecture eq "arm64") {
         determineXcodeSDK();
-        if ($xcodeSDK eq "macosx.internal") {
+        if ($xcodeSDK =~ /osx?\.internal$/) {
```
Perhaps it's better to be explicit that `simulator` is not present in the identifier?

aproskuryakov commented on Tools/Scripts/webkitdirs.pm:
```diff
@@ -599,6 +581,56 @@ sub determineArchitecture
     $architecture = 'arm64' if $architecture =~ /aarch64/i;
 }
 
+sub determineDestination
```
`detemineXcodeDestination`, to make it more obvious that this doesn't apply to other platforms?

aproskuryakov commented on Tools/Scripts/webkitdirs.pm:
```diff
@@ -599,6 +581,56 @@ sub determineArchitecture
+        $destination .= ',variant=Mac Catalyst' if willUseMacCatalystSDK();
+    }
+    
+        
```
Extra newline.

aproskuryakov commented on Tools/Scripts/webkitperl/BuildSubproject.pm:
```diff
@@ -146,6 +146,7 @@ $ENV{'EXPORT_COMPILE_COMMANDS'} = "YES" if $exportCompileCommands;
 
 checkRequiredSystemConfig();
 setConfiguration();
+shouldBuild32Bit();  # consume the --32-bit flag if passed and set internal state
```
Style nit: please make this comment a complete sentence with a period.

aproskuryakov commented on Makefile.shared:
```diff
@@ -31,29 +30,27 @@ endif
 	ifneq (,$(OVERRIDE_SDKROOT))
 		ifneq (default,$(OVERRIDE_SDKROOT))
-			XCODE_OPTIONS := $(XCODE_OPTIONS) SDKROOT=$(OVERRIDE_SDKROOT)
+			BUILD_WEBKIT_OPTIONS := $(BUILD_WEBKIT_OPTIONS) --sdk=$(OVERRIDE_SDKROOT)
```
This is a bit surprising, from a cursory inspection of code `BUILD_WEBKIT_OPTIONS` seems like it's meant to be an input parameter (options passed to `build-webkit`).

emw-apple replied:
Great catch. `BUILD_WEBKIT_OPTIONS` should probably be named something like `BUILD_WEBKIT_ARGS` to indicate this, but there's no point in breaking existing behavior now. I'll move all the flags passed by the Makefile into their own variable.

emw-apple committed:
[webkitdirs] Compute Xcode destination specifier
https://bugs.webkit.org/show_bug.cgi?id=266493
rdar://86340301

Reviewed by Alexey Proskuryakov.

Instead of letting xcodebuild pick an ambiguous destination for us, have
webkitdirs compute a destination string based on the SDK and
architecture(s). This allows us to remove much of the hardcoded
architecture logic in webkitdirs, letting the build system figure out
which architecture(s) to use on its own.

Some nice side effects of this change:

- It is possible to build for a non-Mac platform just by setting
  SDKROOT. No need to override ARCHS.

- When building for macOS or a simulated platform, the default behaviors
  correspond to destinations in Xcode's destination menu. Building with
  no architecture override selects the "My Mac" destination or a
  simulator on the host system. Building with multiple architectures
  selects one of the "Any <platform> Device|Simulator" destinations.
  Aligning the destinations in this way helps keep incremental builds
  functional when switching between Make and Xcode.

To simplify the problem space, remove set-webkit-configuration's ability
to adjust the architecture, and webkitdirs' corresponding knowledge of
"passed" architecture. set-webkit-configuration's architecture flags
have not been updated since Apple's 64-bit transition in 2009
(https://commits.webkit.org/33755@main).

* Makefile.shared: Pass the active SDK and arch to webkitdirs, to
  compute the destination string. Work around shell backticks being
  unable to expand an argument containing spaces with a $(shell ...)
  directive. Since the variable is only expanded once, webkitdirs is
  only called once.

* Tools/Scripts/build-imagediff: No longer need to override architecture.

* Tools/Scripts/set-webkit-configuration: Remove --32-bit and --64-bit
  flags which save state in the build directory. This is an old
  mechanism from before Apple's 64-bit transition, and is not used by
  the present day 32-bit WPE configurations.
(printCurrentSettings):

* Tools/Scripts/webkitdirs.pm:
(determineArchitecture): Don't look for set-webkit-configuration's state
  file.
(determineDestination): Added, computes an xcodebuild destination string
  based on SDK and architecture(s).
(determineXcodeSDKPlatformName): Refactor to allow a "maccatalyst"
  platform name when a SDK is also provided.
(determineXcodeSDK): Special case for "maccatalyst" as a platform name.
(XcodeOptions):
(XcodeOptionString): Enclose arguments containing spaces in quotes, for
  shell substitution.
(setConfiguration):
(destination): Added.
(determinePassedArchitecture): Deleted.
(passedArchitecture): Deleted.
(setArchitecture): Deleted.

* Tools/Scripts/webkitperl/BuildSubproject.pm: Argument parser was
  relying on shouldBuild32Bit() being called early on in evaluation,
  make the call explicit.

Canonical link: https://commits.webkit.org/273153@main
